src/main/scala/cromwell/engine/WorkflowManager.scala

```diff
@@ -25,20 +30,33 @@ trait WorkflowManager {
 
   val backend: Backend
 
-  def generateWorkflow(id: WorkflowId, wdl: WdlSource, inputs: WorkflowInputs): Try[Workflow]
-  def workflowStatus(id: WorkflowId): Option[WorkflowState]
+  def generateWorkflow(id: WorkflowId, wdl: WdlSource, inputs: binding.WorkflowInputs): Try[Workflow]
+
+  def workflowStatus(id: WorkflowId): Option[WorkflowState] = workflowStates.get(id)
+  def workflowOutputs(id: WorkflowId): Future[Option[binding.WorkflowOutputs]]
+  def submitWorkflow(wdl: WdlSource, inputs: binding.WorkflowInputs): Future[Try[WorkflowId]]
+  def updateWorkflowState(workflow: Workflow, state: WorkflowState): Unit
 
-  private val workflowStore = new WriteOnceStore[WorkflowId, Workflow]
+  def idByWorkflow(workflow: Workflow): Option[WorkflowId] = {
+    workflowStore.toMap find {case (k, v) => v == workflow} map {_._1}
+  }
+
+  private val workflowStore = new WriteOnceStore[WorkflowId, Workflow] // This probably doesn't need to be persisted
+  protected val workflowStates = TrieMap.empty[WorkflowId, WorkflowState] // This *should* be persisted
 
   /**
    * Generates a workflow with an ID, inserts it into the store and returns the pair.
    */
-  def submitWorkflow(wdl: WdlSource, inputs: WorkflowInputs): Try[ManagedWorkflow] = {
-    val workflowId = UUID.randomUUID()
-    for {
-      wf <- generateWorkflow(workflowId, wdl, inputs)
-      x <- workflowStore.insert(workflowId, wf) // Come for the side effect, stay for the Try
-    } yield ManagedWorkflow(workflowId, wf)
+  protected def addWorkflow(wdl: WdlSource, inputs: binding.WorkflowInputs): Future[Try[ManagedWorkflow]] = {
```

Comment:
Are the outer curlies necessary if the entire contents are wrapped by a `Future`?  Mostly I ask because I did not put them in the stuff on my branch.  :smile: